File: client-extensions/ruby/verified.rb
require 'digest'
module AgentTrunk
  module Verified
    def self.read_file(client, workspace:, key:, revision:, path:)
      valid_path = path.bytesize.between?(1,512) && !path.match?(/[\\\x00-\x1f\x7f]/) && path.split('/',-1).none? { |s| ['', '.', '..'].include?(s) }
      raise ArgumentError, 'Immutable revision and relative path required' unless revision.match?(/\A[a-f0-9]{64}\z/) && valid_path
      inspected = client.contexts.inspect(trunk_id: workspace, context_key: key, ref: revision)
      raise ArgumentError, 'Revision mismatch' unless inspected.revision.id == revision
      file = inspected.revision.files.find { |f| f.path == path }
      raise ArgumentError, 'Invalid manifest' unless file && file.size.between?(0,1_000_000) && file.sha256.match?(/\A[a-f0-9]{64}\z/)
      bytes = client.contexts.read_file(trunk_id: workspace, context_key: key, resource_path: path, ref: revision)
      raise ArgumentError, 'Integrity check failed' unless bytes.bytesize == file.size && Digest::SHA256.hexdigest(bytes) == file.sha256
      bytes
    end
  end
end

File: client-extensions/python/verified.py
"""Verified immutable file helpers for sync and async generated clients."""
import hashlib
import re


def _manifest(inspected, revision_id, path):
    if not re.fullmatch(r'[a-f0-9]{64}', revision_id) or inspected.revision.id != revision_id:
        raise ValueError('Immutable revision mismatch')
    if not path or len(path.encode()) > 512 or '\\' in path or any(ord(c) < 32 or ord(c) == 127 for c in path) or any(part in ('', '.', '..') for part in path.split('/')):
        raise ValueError('Invalid resource path')
    file = next((file for file in inspected.revision.files if file.path == path), None)
    if file is None or not 0 <= file.size <= 1_000_000 or not re.fullmatch(r'[a-f0-9]{64}', file.sha256):
        raise ValueError('Invalid file manifest')
    return file


def _verify(content, file):
    if len(content) != file.size or hashlib.sha256(content).hexdigest() != file.sha256:
        raise ValueError('Context file integrity check failed')
    return content


def read_verified_file(client, workspace_id, context_key, revision_id, path):
    if not re.fullmatch(r'[a-f0-9]{64}', revision_id):
        raise ValueError('An immutable revision ID is required')
    inspected = client.contexts.inspect(workspace_id, context_key, ref=revision_id)
    file = _manifest(inspected, revision_id, path)
    chunks, size = [], 0
    for chunk in client.contexts.read_file(workspace_id, context_key, path, ref=revision_id):
        size += len(chunk)
        if size > 1_000_000:
            raise ValueError('Context file exceeds size limit')
        chunks.append(chunk)
    return _verify(b''.join(chunks), file)


async def async_read_verified_file(client, workspace_id, context_key, revision_id, path):
    if not re.fullmatch(r'[a-f0-9]{64}', revision_id):
        raise ValueError('An immutable revision ID is required')
    inspected = await client.contexts.inspect(workspace_id, context_key, ref=revision_id)
    file = _manifest(inspected, revision_id, path)
    chunks, size = [], 0
    async for chunk in client.contexts.read_file(workspace_id, context_key, path, ref=revision_id):
        size += len(chunk)
        if size > 1_000_000:
            raise ValueError('Context file exceeds size limit')
        chunks.append(chunk)
    return _verify(b''.join(chunks), file)
